### DISCUSSION TITLE
`Evaluate` question (related to RAD_MAX, but perhaps not only)

### DISCUSSION THREAD

**Discussion by mdpunch** [gammapy/gammapy#4082] [Help] [answered]
Hi all,

(still with gammapy 0.20.1) 

I have a `rad_max` IRF, with just one offset, and energy bins 5/decade from 10GeV to 100TeV (so, 20 bins).

```
>>> rad_max.axes['energy']

MapAxis

        name       : energy    
        unit       : 'TeV'     
        nbins      : 20        
        node type  : edges     
        edges min  : 1.0e-02 TeV
        edges max  : 1.0e+02 TeV
        interp     : log       
```

The data (i.e., the cut applied in each of the 20 bins) is:
```
>>> rad_max.data.T[0]*rad_max.unit
[NaN, NaN, NaN, NaN, NaN, NaN, 1.2017647, 0.96087744, 0.75775545, 0.58407787, 0.45431667, 0.35241865, 0.28885854, 0.23838654, 0.20704, 0.17500642, 0.15643081, 0.14916256, 0.15394908, 0.13236962]∘
```

But, if I try to use evaluate, I get this:
```
>>> rad_max.evaluate(offset=18*u.deg).T[0]
[0, 0, 0, 0, 0, 0, 0, 0.96087744, 0.75775545, 0.58407787, 0.45431667, 0.35241865, 0.28885854, 0.23838654, 0.20704, 0.17500642, 0.15643081, 0.14916256, 0.15394908, 0.13236962]∘
```
... mostly fine, except the element [6] changes from 1.2017647 to 0 !!

Checking further:
```
>>> rad_max.axes['energy'].center[6]
0.19952623TeV
>>> rad_max.evaluate(offset=18*u.deg,energy=0.19952623*u.TeV)
0∘
rad_max.evaluate(offset=18*u.deg,energy=0.2*u.TeV)
1.2005242∘
```

In short, I did not expect `evaluate` to give me a zero value for `rad_max` just below the centre of the bin, which it happens to fall below because of rounding errors.

Of course, this only happens because I have `NaN` in the lower bins.  If the earlier bins have zero (0) the interpolation is fine, but having zero gives errors later.

So, is it possible to change the behaviour of `evaluate` such that if there is an `NaN` in the next bin, then the interpolation sets the value to the value of the bin rather than setting to 0?

This would avoid unstable behaviour due to rounding errors...

  Good Luck,
    Michael.

**Comment by adonath** [comment, marked as answer]
Thanks @mdpunch! What happens if you use nearest neighbor interpolation instead:

```python3
rad_max.evaluate(offset=18*u.deg, method="nearest")
```

**Comment by mdpunch** [comment]
Nice!
'nearest' is what I was looking for:
![image](https://user-images.githubusercontent.com/17853426/190679795-95ab764a-48d3-4136-b0ce-56780fa310b3.png)
Thanks!

**Reply by adonath** [reply]
Glad I could help!
